compiler/rustc_mir/src/transform/promote_consts.rs

```diff
@@ -626,9 +649,28 @@ impl<'tcx> Validator<'_, 'tcx> {
                     ));
                     return Err(Unpromotable);
                 }
+                if !self.explicit && !lhs_ty.is_integral() {
+                    // Only allow this on integers (not on floats).
+                    return Err(Unpromotable);
+                }
 
                 match op {
-                    // FIXME: reject operations that can fail -- namely, division and modulo.
+                    BinOp::Div | BinOp::Rem => {
+                        if !self.explicit {
+                            // The RHS must be a non-zero const.
+                            let const_val = match rhs {
+                                Operand::Constant(c) => {
+                                    c.literal.try_eval_bits(self.tcx, self.param_env, lhs_ty)
+                                }
+                                _ => None
+                            };
+                            match const_val {
+                                Some(x) if x != 0 => {} // okay
```

Comment:
This is very pedantic and not related to PR logic at all, but isn't this a great situation for

```suggestion
                                Some(1..) => {} // okay
```

Reply:
This is `x > 0`, not `x != 0`, which is only equivalent for unsigned types, and it is not locally obvious that `x` is unsigned. So I think I'd prefer to stick to `!= 0`.